### PR TITLE
Token caching and validation

### PR DESCRIPTION
**Pull request by sastaachar**
Adding Caching for token 
Added tokenised fetch
Using tokenised fetch for many apis
Had to move few dependency to separate files to avoid circular dependencey

### CODE REVIEW

File: src/authToken.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+const validateAuthToken = async (embedConfig :EmbedConfig, authToken: string): Promise<boolean> => {
+    try {
+        const isTokenValid = await verifyTokenService(embedConfig.thoughtSpotHost, authToken);
+        if (isTokenValid) return true;
+
+        if (cachedAuthToken && cachedAuthToken === authToken) {
+            // eslint-disable-next-line no-alert
+            alert(DUPLICATE_TOKEN_ERR);
+            throw new Error(DUPLICATE_TOKEN_ERR);
```
Comment (ashubham) [inline src/authToken.ts:47]: what the point of throwing errors here, if we are anyways catch it in the same block.

File: src/utils/graphql/answerService/answerService.ts [hunk 2]
```diff
@@ -103,7 +104,7 @@ export class AnswerService {
      */
     public async fetchCSVBlob(userLocale = 'en-us', includeInfo = false): Promise<Response> {
         const fetchUrl = this.getFetchCSVBlobUrl(userLocale, includeInfo);
-        return fetch(fetchUrl, {
+        return tokenizedFetch(fetchUrl, {
```
Comment (ashubham) [inline src/utils/graphql/answerService/answerService.ts:107]: Are there other places we need to use tokenizedFetch ?

File: src/tokenizedFetch.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import { getAuthenticationToken } from './authToken';
+import { getEmbedConfig } from './embed/base';
+
+export const tokenizedFetch: typeof fetch = async (input, init) : Promise<Response> => {
+    const req = new Request(input, init);
+    const embedConfig = getEmbedConfig();
+    const authToken = await getAuthenticationToken(embedConfig);
```
Comment (ashubham) [inline src/tokenizedFetch.ts:7]: What if the user is not using TrustedAuthCookieless ?

**Review by ashubham — CHANGES_REQUESTED** [review]

File: src/utils/graphql/answerService/answerService.ts (hunk 2 above)
Comment (sastaachar) [reply src/utils/graphql/answerService/answerService.ts:107]: yes , execute and import TML , 

cc: @adityamittal3107 @Nirmay15  please do suggest if we need this anywhere else

File: src/auth.ts [hunk 4]
```diff
@@ -363,13 +320,12 @@ export const doCookielessTokenAuth = async (embedConfig: EmbedConfig): Promise<b
     }
     try {
         const authToken = await getAuthenticationToken(embedConfig);
-        const response = await verifyTokenService(embedConfig.thoughtSpotHost, authToken);
-        if (!response.ok) return false;
-    } catch (e) {
-        return false;
+        if (authToken) return true;
```
Comment (abdullasulaiman) [inline src/auth.ts:323]: let isAuthenticated = false;
if (authToken) { isAuthenticated = true };

File: src/authToken.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    if (cachedAuthToken) {
+        try {
+            const isCachedTokenStillValid = await validateAuthToken(
+                embedConfig,
+                cachedAuthToken,
+                true,
+            );
+            if (isCachedTokenStillValid) return cachedAuthToken;
+        } catch {
+            // Continue to get a new token if validation fails
```
Comment (abdullasulaiman) [inline src/authToken.ts:22]: Look at this block to make for debugging purpose.

**Review by abdullasulaiman — CHANGES_REQUESTED** [review]

**Review by abdullasulaiman — APPROVED** [review]
LGTM

**Review by ashubham — APPROVED** [review]

**Pull request merged**